n-api: handle weak no-finalizer refs correctly

gabrielschulhof opened the pull request:
When deleting a weak reference that has no finalizer we must not defer
deletion until the non-existent finalizer gets called.

Fixes: https://github.com/nodejs/node/issues/34731

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

addaleax commented on src/js_native_api_v8.cc:
```diff
@@ -242,6 +242,8 @@ class RefBase : protected Finalizer, RefTracker {
   static inline void Delete(RefBase* reference) {
     reference->Unlink();
     if ((reference->RefCount() != 0) ||
+        (reference->RefCount() == 0 &&
```
`RefCount()` is unsigned, so `reference->RefCount() == 0` is always true here, right?

gabrielschulhof replied:
That's true. This can be simplified to the second half of the `&&`.

gabrielschulhof commented:
@addaleax I removed the always-true condition and I also updated the comment above the function to include the new condition (i.e. that we want to delete when the ref is weak but has no finalizer).

gabrielschulhof commented:
Added Signed-off-by header.

gabrielschulhof commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/32854/

gabrielschulhof commented:
Hmmm ... CI is failing, but the logs are truncated, so I can't see what the failure was.

gabrielschulhof commented:
Resumed as: https://ci.nodejs.org/job/node-test-pull-request/32860/
